### PR TITLE
Add exp_pauli quantum instruction to the runtime and ASTBridge

### PR DESCRIPTION
**Pull request by amccaskey**
Enable the following in the runtime and the bridge: 
```cpp
auto kernel = [](double theta) __qpu__ {
  cudaq::qreg q(4);
  x(q[0]);
  x(q[1]);
  exp_pauli(theta, q, "XXXY");
};

auto kernel2 = [](double theta, const char * pauli) __qpu__ {
   cudaq::qreg q(4);
   x(q[0]);
   x(q[1]);
   exp_pauli(theta, q, pauli);
 };

auto kernel3 = [](double theta) __qpu__ {
  cudaq::qreg q(4);
  x(q[0]);
  x(q[1]);
  exp_pauli(theta, "XXXY", q[0], q[1], q[2], q[3]);
};

auto [kernel, theta] = cudaq::make_kernel<double>();
auto qubits = kernel.qalloc(4);
kernel.x(qubits[0]);
kernel.x(qubits[1]);
kernel.exp_pauli(theta, qubits, "XXXY");

auto [kernel, theta] = cudaq::make_kernel<double>();
auto qubits = kernel.qalloc(4);
kernel.x(qubits[0]);
kernel.x(qubits[1]);
kernel.exp_pauli(theta, "XXXY", qubits[0], qubits[1], qubits[2], qubits[3]);
```
```python
kernel, theta = cudaq.make_kernel(float)
qubits = kernel.qalloc(4)
kernel.x(qubits[0])
kernel.x(qubits[1])
kernel.exp_pauli(theta, qubits, "XXXY")
```

Also: 

Add a `cc.string` type, a `cc.string_literal` op, and updates to lower to all this LLVM.

### CODE REVIEW

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td [hunk 1]
```diff
@@ -754,6 +754,23 @@ class TwoTargetOp<string mnemonic, list<Trait> traits = []> :
 // Quantum operators (gates)
 //===----------------------------------------------------------------------===//
 
+def ExpPauliOp : QuakeOp<"exp_pauli", []> {
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:757]: The new ops and types should be added to the roundtrip.qke test.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td [hunk 2]
```diff
@@ -754,6 +754,23 @@ class TwoTargetOp<string mnemonic, list<Trait> traits = []> :
...
+  let summary = "General Pauli tensor product rotation";
+  let description = [{
+    This operation affects a general Pauli tensor product rotation on 
+    the input qubits. The number of Pauli characters in the input Pauli word 
+    string must equal the number of qubits in the veq. Mathematically, this operation 
+    applies exp(-i theta P) where P is a general Pauli tensor product.
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:763]: From this description, I'm thinking `std::string` is maybe a bit heavyweight for this. We probably want the programming interface to take an actual string literal and add the template constraints to make sure it has the correct size per the number of arguments, etc.

Then at the quake level, we can just use a type of `!cc.array<i8 x `_n_`>` for a C-string literal.

File: lib/Optimizer/CodeGen/LowerToQIR.cpp [hunk 3]
```diff
@@ -1214,6 +1260,47 @@ class StdvecSizeOpPattern
...
+    StringRef stringLiteral = stringLiteralOp.getStringLiteral();
+
+    // Write to the module body
+    auto insertPoint = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointToStart(parentModule.getBody());
+
+    // Create the register name global
+    auto builder = cudaq::IRBuilder::atBlockEnd(parentModule.getBody());
+    auto slGlobal =
+        builder.genCStringLiteralAppendNul(loc, parentModule, stringLiteral);
```
Comment (schweitzpgi) [inline lib/Optimizer/CodeGen/LowerToQIR.cpp:1286]: Cool. I learned something. Back in the day with FIR, I had to roll my own method for this. :)

Comment (amccaskey) [reply lib/Optimizer/CodeGen/LowerToQIR.cpp:1286]: I copied this from code you had written 😄

Comment (schweitzpgi) [reply lib/Optimizer/CodeGen/LowerToQIR.cpp:1286]: I promptly forgot about this it appears. 😄

**Comment by poojarao8** [comment]
This looks great and it'd be really useful in several ongoing projects with partners. Is there any way we can bump this up? @bettinaheim

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td (hunk 2 above)
Comment (amccaskey) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:795]: I think the counter-argument (and there is an example in the tests for this) is the case where the user wants to pass the string as a kernel argument.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td (hunk 1 above)
Comment (amccaskey) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:789]: Will do.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td (hunk 2 above)
Comment (schweitzpgi) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td]: The counterargument is `std::string` is not a literal nor the type of a literal. It's a C++ class (actually a template specialization of a parametric polymorphic class) with allocators, data members, a number of methods for doing things with strings, iterators, hashing, etc. Supporting all that "extra stuff" doesn't seem necessary. Yet the implication of adding a `std::string` type to quake is that quake will possibly support those features.

**Comment by amccaskey** [comment]
documenting this, for some reason on CI, we get a function call from `libnvqir-custatevec-fp32`?
```bash

Thread 1 "python3.11" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
#0  0x0000000000000000 in ?? ()
#1  0x00007f94325425eb in std::_Function_handler<void (bool), nvqir::CircuitSimulator::applyExpPauli(double, std::vector<unsigned long, std::allocator<unsigned long> > const&, cudaq::spin_op const&)::{lambda(cudaq::pauli, unsigned long)#1}::operator()(cudaq::pauli, unsigned long) const::{lambda(bool)#2}>::_M_invoke(std::_Any_data const&, bool&&) ()
   from /root/.local/lib/python3.11/site-packages/cudaq/../lib/libnvqir-custatevec-fp32.so
#2  0x00007f94324c2d4e in ?? ()
   from /root/.local/lib/python3.11/site-packages/cudaq/../lib/libnvqir-qpp.so
#3  0x00007f945faffbef in __quantum__qis__exp_pauli ()
   from /root/.local/lib/python3.11/site-packages/cudaq/../lib/libnvqir.so
#4  0x00007f946105f0b4 in __nvqpp__mlirgen____nvqppBuilderKernel_108506462586
    ()
#5  0x00007f946105f1b1 in __nvqppBuilderKernel_108506462586.thunk ()
#6  0x00007f945fce3e33 in ?? ()
   from /root/.local/lib/python3.11/site-packages/cudaq/../lib/libcudaq-platform-default.so
#7  0x00007f945fb532d5 in altLaunchKernel ()
```

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td (hunk 2 above)
Comment (amccaskey) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td]: I see your point - I will update to remove `cc.string` and have `cc.string_literal` return a `cc.array<i8*N>` and input `std::string` kernel arguments to `cc.ptr<i8>`.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td [hunk 4]
```diff
@@ -786,6 +786,23 @@ class TwoTargetOp<string mnemonic, list<Trait> traits = []> :
...
+def ExpPauliOp : QuakeOp<"exp_pauli", []> {
+  let summary = "General Pauli tensor product rotation";
+  let description = [{
+    This operation affects a general Pauli tensor product rotation on 
+    the input qubits. The number of Pauli characters in the input Pauli word 
+    string must equal the number of qubits in the veq. Mathematically, this operation 
+    applies exp(i theta P) where P is a general Pauli tensor product. 
+  }];
+  
+  let arguments = (ins AnyFloat:$parameter, VeqType:$qubits, cc_PointerType:$pauli);
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:798]: Noting that this should be extended to allow wires, etc.

File: runtime/cudaq/builder/kernel_builder.cpp [hunk 5]
```diff
@@ -163,6 +163,40 @@ bool isArgStdVec(std::vector<QuakeValue> &args, std::size_t idx) {
...
+                             "type as first argument.");
+  cudaq::info("kernel_builder apply exp_pauli {}", pauliWord);
+
+  auto strLitTy = cc::PointerType::get(cc::ArrayType::get(
+      builder.getContext(), builder.getI8Type(), pauliWord.size() + 1));
+  Value stringLiteral = builder.create<cc::CreateStringLiteralOp>(
+      strLitTy, builder.getStringAttr(pauliWord));
+  SmallVector<Value> args{thetaVal, qubitsVal, stringLiteral};
+  builder.create<quake::ExpPauliOp>(TypeRange{}, args);
+  return;
```
Comment (schweitzpgi) [inline runtime/cudaq/builder/kernel_builder.cpp:197]: nit: not needed
```suggestion
```

File: test/AST-Quake/exp-pauli-1.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+int main() {
+  auto kernel = [](double theta) __qpu__ {
+    cudaq::qreg q(4);
+    x(q[0]);
+    x(q[1]);
+    exp_pauli(theta, q, "XXXY");
+  };
+}
+
+
```
Comment (schweitzpgi) [inline test/AST-Quake/exp-pauli-1.cpp:22]:
```suggestion

```

File: test/AST-Quake/exp-pauli-2.cpp [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  auto kernel2 = [](double theta, const char *pauli) __qpu__ {
+    cudaq::qreg q(4);
+    x(q[0]);
+    x(q[1]);
+    exp_pauli(theta, q, pauli);
+  };
+}
+
+
+
```
Comment (schweitzpgi) [inline test/AST-Quake/exp-pauli-2.cpp:23]:
```suggestion

```

File: unittests/CMakeLists.txt [hunk 8]
```diff
@@ -64,7 +64,7 @@ macro (create_tests_with_backend NVQIR_BACKEND EXTRA_BACKEND_TESTER)
     nvqir-${NVQIR_BACKEND} nvqir  
     cudaq fmt::fmt-header-only
     cudaq-platform-default
-    cudaq-builder
+    cudaq-builder
```
Comment (schweitzpgi) [inline unittests/CMakeLists.txt:67]:
```suggestion
    cudaq-builder
```

**Review by schweitzpgi — COMMENTED** [review]
nits.

Looks good to me, but I helped write the code. :) @1tnguyen can you take a look?

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td (hunk 4 above)
Comment (amccaskey) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:798]: I will create a tracker for it. I will need your help on it.

File: runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu [hunk 9]
```diff
@@ -363,6 +363,34 @@ public:
...
+      else
+        paulis.push_back(custatevecPauli_t::CUSTATEVEC_PAULI_Z);
+
+      targets.push_back(qubits[i]);
+    });
+
+    custatevecApplyPauliRotation(
+        handle, deviceStateVector, cuStateVecCudaDataType, nQubitsAllocated,
+        theta, paulis.data(), targets.data(), targets.size(), controls.data(),
+        nullptr, controls.size());
```
Comment (1tnguyen) [inline runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu:391]:
```suggestion
    HANDLE_ERROR(custatevecApplyPauliRotation(
        handle, deviceStateVector, cuStateVecCudaDataType, nQubitsAllocated,
        theta, paulis.data(), targets.data(), targets.size(), controls.data(),
        nullptr, controls.size()));
```

File: runtime/nvqir/qpp/QppCircuitSimulator.cpp [hunk 10]
```diff
@@ -181,6 +181,49 @@ class QppCircuitSimulator : public nvqir::CircuitSimulatorBase<double> {
...
+    for (auto &[i, j] : toReverse)
+      x({i}, j);
+
+    if (!basisChange.empty()) {
+      std::reverse(basisChange.begin(), basisChange.end());
+      for (auto &basis : basisChange)
+        basis(true);
+    }
+  }
+
```
Comment (1tnguyen) [inline runtime/nvqir/qpp/QppCircuitSimulator.cpp:226]: Question only: I wonder if this should be the base implementation rather than throwing.

**Review by 1tnguyen — APPROVED** [review]
LGTM 👍

File: runtime/nvqir/qpp/QppCircuitSimulator.cpp (hunk 10 above)
Comment (amccaskey) [reply runtime/nvqir/qpp/QppCircuitSimulator.cpp:226]: It should, but this was the workaround for the CI bug.

**Pull request merged**